### PR TITLE
Custom Secret

### PR DESCRIPTION
**Pull request by dmathieu**
This adds custom secret : 

With a static string (no variables can be used here)

```
attr_secure :my_attribute, :secret => "..."
```

With a lambda (object is the current instance of the atribute's object).

```
attr_secure :my_attribute, :secret => lambda {|object| object.user.secret }
```

### CODE REVIEW

**Comment by dmathieu** [comment]
@neilmiddleton: I don't really see the point for string secrets.
A lambda one is interesting as it allows having differents secrets depending of the instance.
However, the only point I see for strings would be to use an other environment variable.

File: lib/attr_secure.rb [hunk 1]
```diff
@@ -20,19 +20,29 @@ module AttrSecure
...
+
     define_method("#{attribute}=") do |value|
-      encrypted_value = encryption_class.encrypt(value.nil? ? nil : value)
+      options[:encryption_class].object = self
+      encrypted_value = options[:encryption_class].encrypt(value.nil? ? nil : value, options[:secret])
       self.class.attr_secure_adapter.write_attribute self, attribute, encrypted_value
     end
 
     define_method("#{attribute}") do
+      options[:encryption_class].object = self
```
Comment (dmathieu) [inline lib/attr_secure.rb]: This is not cool. But we need to know the current instance in `encryption_class` in order to send it to the lambda.

**Comment by neilmiddleton** [comment]
Fair point with the lambda vs strings - I'm happy with that.

File: lib/attr_secure.rb (hunk 1 above)
Comment (neilmiddleton) [inline lib/attr_secure.rb]: What do you think the potential problems might be?

Comment (dmathieu) [inline lib/attr_secure.rb]: It's not about a problem. It's about readiness.
I have an idea for a solution though. I'll look into it soon.

**Comment by dmathieu** [comment]
@neilmiddleton: I've just split the secret retrieval into it's own object.
I've also (re)refactored the tests to not make them interdependent.
If that's fine with you, this can be merged.

**Comment by neilmiddleton** [comment]
Awesome.  :thumbsup:

**Pull request merged**